go/extractor/extractor.go

```diff
@@ -132,10 +133,33 @@ func ExtractWithFlags(buildFlags []string, patterns []string) error {
 
 	pkgsNotFound := make([]string, 0, len(pkgs))
 
+	// Build a map from package paths to their longest IDs--
+	// in the context of a `go test -c` compilation, we will see the same package more than
+	// once, with IDs like "abc.com/pkgname [abc.com/pkgname.test]" to distinguish the version
+	// that contains and is used by test code.
+	// For our purposes it is simplest to just ignore the non-test version, since the test
+	// version seems to be a superset of it.
+	longestPackageIds := make(map[string]string)
+	packages.Visit(pkgs, nil, func(pkg *packages.Package) {
+		if shortestID, present := longestPackageIds[pkg.PkgPath]; present {
+			if len(pkg.ID) > len(shortestID) {
+				longestPackageIds[pkg.PkgPath] = pkg.ID
+			}
+		} else {
+			longestPackageIds[pkg.PkgPath] = pkg.ID
+		}
+	})
+
 	// Do a post-order traversal and extract the package scope of each package
 	packages.Visit(pkgs, nil, func(pkg *packages.Package) {
 		log.Printf("Processing package %s.", pkg.PkgPath)
 
+		// If this is a variant of a package that also occurs with a longer ID, skip it.
+		if pkg.ID != longestPackageIds[pkg.PkgPath] {
+			log.Printf("Skipping variant of package %s with ID %s.", pkg.PkgPath, pkg.ID)
```

Comment:
For a large project with a lot of packages, I imagine that this creates a lot of noise in the extractor log. Should we try and reduce this by e.g. having a list of all the skipped package variants and printing the first N-many in one go as we have done elsewhere?

Reply:
For now I decline; this is a niche extraction mode and I would prefer that we're verbose for clarity. We're also already logging a line per package seen in various places, so I don't expect a major blowup.

Reply:
Note that `JoinTruncatedList` in our `util` package may do most of the work for you already.